### PR TITLE
Added Mozilla Firefox

### PR DESCRIPTION
**Pull request by KOLANICH**
- [x] I followed the contributing guidelines: https://github.com/github/explore/blob/master/CONTRIBUTING.md

I am:
  - [ ] Suggesting edits to an existing Topic page
  - [x] Curating a new Topic page

***********EDITING AN EXISTING TOPIC PAGE************

I'm suggesting these edits to an existing topic:
- [ ] Image (and my file is `*.png`, square, dimensions 288x288)
- [ ] Content (and my changes are in `index.md`)

Please explain why these changes are necessary:

************CURATING A NEW TOPIC PAGE************

- [x] I've formatted my changes as a new folder directory, named for the topic as it appears in the URL on GitHub (e.g. `https://github.com/topics/[NAME]`)
- [ ] My folder contains a `*.png` image (if applicable) and `index.md`
- [x] All required fields in my `index.md` conform to the Style Guide and API docs: https://github.com/github/explore/tree/master/docs

Please explain why you think this Topic page should be curated:

It's a platform for extensions, both XPCOM and WebExt.

### CODE REVIEW

File: topics/firefox/index.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+display_name: Mozilla Firefox
+github_url: https://github.com/mozilla
+released: '2002-09-23'
+short_description: An advanced free open-source web browser
+topic: firefox
+url: https://mozilla.org/firefox
+wikipedia_url: https://en.wikipedia.org/wiki/Firefox
+related: WebExtension
+---
+A web browser by Mozilla which once was the most hackable and had the most powerful extensions.
```
Comment (MikeMcQuaid) [inline topics/firefox/index.md]: `A web browser by Mozilla which is hackable and has powerful extensions.

File: topics/firefox/index.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+---
+aliases: firefox, seamonkey, waterfox, palemoon, icecat, iceweasel, torbrowser
+created_by: Mozilla
+display_name: Mozilla Firefox
+github_url: https://github.com/mozilla
+released: '2002-09-23'
+short_description: An advanced free open-source web browser
```
Comment (MikeMcQuaid) [inline topics/firefox/index.md]: `An advanced, free, open-source web browser`

**Comment by MikeMcQuaid** [comment]
@KOLANICH Thanks for the topic! Can you fix up the failing tests: https://travis-ci.org/github/explore/builds/300437360?utm_source=github_status&utm_medium=notification

File: topics/firefox/index.md (hunk 1 above)
Comment (KOLANICH) [reply topics/firefox/index.md]: It's not anymore.

File: topics/firefox/index.md (hunk 2 above)
Comment (KOLANICH) [reply topics/firefox/index.md]: OK

File: topics/firefox/index.md (hunk 1 above)
Comment (MikeMcQuaid) [reply topics/firefox/index.md]: @KOLANICH Even if that's the case it's a bit negative/passive aggressive to reference this in the past tense so can you try to reword this to talk about the current state of Firefox rather than what it used to be. How about `Mozilla Firefox is a fast, cross-platform web browser developed by the Mozilla Foundation.`?

Comment (MikeMcQuaid) [reply topics/firefox/index.md]: @KOLANICH Can you update this so there is no past tense in the description, please? Thanks!

Comment (KOLANICH) [reply topics/firefox/index.md]: >How about Mozilla Firefox is a fast, cross-platform web browser developed by the Mozilla Foundation.?

Not quite good. There are some legacy (now - obsolete and deprecated) extensions on GitHub. This tag is needed only for them, because WebExtension API is crossbrowser. So I guess we need to put there some clue why these extensions don't work anymore and won't work. Maybe we don't need this tag at all, since the API is deprecated and dropped, these extensions have no future. Al least if someone not forked Firefox and started maintaining the fork, fixing xpcom to becompatible with the latest mozilla code base, building builds and resolving merge conflicts.

File: topics/firefox/index.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+---
+aliases: seamonkey, waterfox, palemoon, icecat, iceweasel, torbrowser
```
Comment (cheshire137) [inline topics/firefox/index.md]: `torbrowser` seems more like a related topic than an alias. Would you move it to line 11?

File: topics/firefox/index.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+created_by: Mozilla
+display_name: Mozilla Firefox
+github_url: https://github.com/mozilla
+released: September 23, 2002
+short_description: An advanced, free, open-source web browser.
```
Comment (cheshire137) [inline topics/firefox/index.md]: Could you rephrase this as a full sentence and remove the hyphen from "open-source"? Maybe "Firefox is an advanced, free, open source web browser."

File: topics/firefox/index.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+topic: firefox
+url: https://mozilla.org/firefox
+wikipedia_url: https://en.wikipedia.org/wiki/Firefox
+related: WebExtension
```
Comment (cheshire137) [inline topics/firefox/index.md]: https://github.com/topics/WebExtension 404s, but https://github.com/topics/web-extension works. Can you change this to `web-extension`?

File: topics/firefox/index.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+---
+A web browser by [Mozilla](https://mozilla.org/) which once [was](https://blog.mozilla.org/addons/2015/08/21/the-future-of-developing-firefox-add-ons/) the most hackable and had the most powerful extensions.
```
Comment (cheshire137) [inline topics/firefox/index.md]: It feels strange making just the word "was" a link.

Can you elaborate more on Firefox? Rather than describing what it used to be, how about comparing it with other popular browsers? Why choose Firefox over Chrome, IE, Safari, for example?

File: topics/firefox/index.md (hunk 3 above)
Comment (KOLANICH) [reply topics/firefox/index.md]: torbrowser is just a tweaked firefox esr, Firefox extensions can be installed to it and its extensions (which are in fact Firefox extensions) can be installed on old (<57, where their support was dropped) Firefox.

File: topics/firefox/index.md (hunk 5 above)
Comment (KOLANICH) [reply topics/firefox/index.md]: I guess that ```web-extension``` should be an alias to ```WebExtension``` since `WebExtension` is an official name.

File: topics/firefox/index.md (hunk 6 above)
Comment (KOLANICH) [reply topics/firefox/index.md]: >Rather than describing what it used to be, how about comparing it with other popular browsers? 

For ordinary people the best browser is the one which is already installed on their PC, disrespect to customizeability (ordinary people don't need it), privacy (ordinary people don't care about it) and performance (modern browsers are fast enough for all the apps any web browser is fast enough, so performance differences are negligible and don't affect UX). Firefox was not a browser for ordinary people, it was the browser for IT maniacs like those who have over 50 addons (I had 56 and I really miss them) made by IT maniacs, but it doesn't mean that it is unsuitable for ordinary people. But IT maniacs often have cognitive biases to think that everyone is an IT maniac, so they have lost the marketing war with Google, which targeted ordinary people with aggressive promotion (Chrome was advertised over TV, chromium-based browsers were and are bundled into installers). Now they are trying to save the situation by cutting features for maniacs, but they have already lost. People don't need their Rust-powered security and multithreaded speed, maniacs don't need the browser without mighty addons.

Comment (cheshire137) [reply topics/firefox/index.md]: Hmm, I'm just trying to get a longer, more informative description. What makes Firefox special?

Comment (malandrina) [reply topics/firefox/index.md]: @KOLANICH 56 addons, wow! It sounds like you were (are?) a Firefox power user.

I took a look at the [Android topic description](https://github.com/github/explore/blob/33196c3d161abc6053e9a56b84d31859f9c00ccf/topics/android/index.md) for inspiration and I feel like the Firefox description would benefit from the addition of some of the following details:

- an indicator of the maturity of the project, e.g. the date of the initial release
- its dependencies
- a concise summary of how it's different from other browsers

@KOLANICH, what do you think about taking some of what you said in https://github.com/github/explore/pull/102/files#r152068729 and simplifying it for inclusion in the topics description? For example:

> Firefox was not a browser for ordinary people, it was the browser for IT maniacs like those who have over 50 addons (I had 56 and I really miss them) made by IT maniacs, but it doesn't mean that it is unsuitable for ordinary people

... could become something as simple as "Firefox supports customization with add-ons".

Comment (KOLANICH) [reply topics/firefox/index.md]: >"Firefox supports customization with add-ons".

It says nothing. Chrome, Chromium-based Opera, Chromium-based Yandex browser, Chromium-based Mail.ru Amigo, Chromium-based Vivaldi, Microsoft Edge, Safary - all of them support customization with WebExtension addons. In fact Chrome supports more addons now than Firefox - WebExtension API is based on Chromium API, and Chrome Web Store is full of addons, only part of them uses the subset of API implemented bot in Firefox and Chromium. What was the distinctive feature of Firefox was the full access to the part of browser's internals written in JS, XUL and HTML. This allowed creation of much powerful addons, some of them even used native shared libraries (using `ctypes` API) and had interacted with operation system. For example it was possible to send and receive network packets, this allowed to create a BitTorrent and an IRC clients as extensions. Or there was an addon taking screenshots. Or there were addons implementing experimental hardware-accelerated (but insecure, these addons were targeted to developers who want start hacking with the new API) API such as WebCL (OpenCL for Web) and RiverTrail (a JavaScript to OpenCL compiler). There were addons adding features into browser interface, for example an addon adding a bar in the bottom to show progress of downloads. Or Tab Mix Plus which allowed customization of tabbar (it allowed to customize tabs behavior, lock some tabs, abd what is the most important to me - it allowed to put tabs in multiple rows). When Mozilla have started transforming browser interface, Classic Theme Restorer addon appeared which allowed to restore old good interface and old good search bar. When Mozilla have removed an awesome Panorama feature it was incapsulated into an addon. There were some privacy protecting addons like NoScript implemented reinforcing the browser against some attacks like clickjacking and XSS.

I mean that webextension addons will never be able to do the same what non-webext addons were able to do. But that had cost - xpcom addons often got broken after a firefox release. After some times they either got fixed by their dev, or forked and fixed by someone else. The addons could violate privacy and security, but AMO is premoderated, all the code was checked by humans, in order to publish there a dev must provide the unobfuscated source code of the addon. If a dev used obfuscation, it must provide an obfuscator so the moderators are able to check that the unobfuscated code corresponds to the obfuscated one.

In 57 Mozilla have disabled XPCOM API for addons for non-nightly versions instead of just changing the internals and allowing the devs to adapt their addons to changing internals. This means that there will be no more so powerful addons we loved Firefox for. They have a process to request a new API but many devs have claimed that they have asked, but their proposals were rejected.

Comment (malandrina) [reply topics/firefox/index.md]: @KOLANICH, it sounds like you think my suggestion to add "Firefox supports customization with add-ons" to the Firefox topic description is reductive and overly simplistic. Is that correct?

To clarify my original intent, I made the suggestion to add "Firefox supports customization with add-ons" to the topic description because it seemed like you were struggling to come up with a concise summary of what sets Firefox apart from other browsers.

I'm still interested to hear your thoughts on my question in https://github.com/github/explore/pull/102#discussion_r152095068:

> what do you think about taking some of what you said in https://github.com/github/explore/pull/102/files#r152068729 and simplifying it for inclusion in the topics description?

Comment (MikeMcQuaid) [reply topics/firefox/index.md]: > It says nothing.

@KOLANICH you're being rude at this point. Please change the text to that suggested by one of us or close out this PR, thanks.

Comment (KOLANICH) [reply topics/firefox/index.md]: >@KOLANICH, it sounds like you think my suggestion to add "Firefox supports customization with add-ons" to the Firefox topic description is reductive and overly simplistic. Is that correct?

It is correct. There are some legacy addons repos on GitHub. Some legacy addons are still in active development, for example μBlock Origin (it is still possible to install and use them on pre57 firefox, for example ESR). But for now it's not very distinct. It's still the most customizeable browser (a tonns of prefs in `about:config`), but its killer feature was removed.

>I'm still interested to hear your thoughts on my question in #102 (comment):

I guess no, it was just an informal and too emotional introduction to the situation around Firefox for the ones who is not very familiar with it. I guess such text may be good for comments, or for an article in some informal encyclopedia, such as Dramatica, but not for Wikipedia or this repo. I guess I can add something to the file, but it will hardly be based on that comment.
